File: apina-spring/src/test/java/fi/evident/apina/spring/SpringUriTemplateParserTest.java
package fi.evident.apina.spring;

import fi.evident.apina.model.URITemplate;
import org.junit.Test;

import static fi.evident.apina.spring.SpringUriTemplateParser.parseUriTemplate;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class SpringUriTemplateParserTest {

    @Test
    public void templateWithoutVariables() {
        assertThat(parseUriTemplate(""), is(new URITemplate("")));
        assertThat(parseUriTemplate("foo"), is(new URITemplate("foo")));
        assertThat(parseUriTemplate("/foo"), is(new URITemplate("/foo")));
    }

    @Test
    public void templateWithSimpleVariables() {
        assertThat(parseUriTemplate("foo/{bar}/{baz}"), is(new URITemplate("foo/{bar}/{baz}")));
    }

    @Test
    public void templateWithVariableConstraints() {
        assertThat(parseUriTemplate("foo/{bar:\\d+}/{baz:[ab]}"), is(new URITemplate("foo/{bar}/{baz}")));
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/spring/SpringUriTemplateParserTest.kt
package fi.evident.apina.spring

import fi.evident.apina.model.URITemplate
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class SpringUriTemplateParserTest {

    @Test
    fun `template without variables`() {
        assertEquals(URITemplate(""), parseSpringUriTemplate(""))
        assertEquals(URITemplate("foo"), parseSpringUriTemplate("foo"))
        assertEquals(URITemplate("/foo"), parseSpringUriTemplate("/foo"))
    }

    @Test
    fun `template with simple variables`() {
        assertEquals(URITemplate("foo/{bar}/{baz}"), parseSpringUriTemplate("foo/{bar}/{baz}"))
    }

    @Test
    fun `template with variable constraints`() {
        assertEquals(URITemplate("foo/{bar}/{baz}"), parseSpringUriTemplate("foo/{bar:\\d+}/{baz:[ab]}"))
    }

    @Test
    fun `templates with closing braces inside constraints`() {
        assertEquals(URITemplate("/{id}/foo"), parseSpringUriTemplate("/{id:[a-z]{16}[0-9]{4}}/foo"))
    }

    @Test
    fun `templates with escaped braces`() {
        assertEquals(URITemplate("/{id}/foo"), parseSpringUriTemplate("/{id:\\}}/foo"))
    }
}
